### ISSUE TITLE
Support for maximum event age

### ISSUE THREAD

**Issue by SanjayS5** [serverless/serverless#7678]
# Feature Proposal

Maximum event age setting

## Description

AWS introduces a [maximum event age](https://aws.amazon.com/about-aws/whats-new/2019/11/aws-lambda-supports-max-retry-attempts-event-age-asynchronous-invocations/) setting for lambdas in November 2019. It would be great if Serverless introduced support for this setting.

**Comment by medikoo** [comment]
@SanjayS5 great thanks for proposal. We're definitely open for PR that implements it. I take it should be referenced via `maximumEventAge` property on function config

**Comment by AhmedFat7y** [comment]
@medikoo In CF, this is part of the destinations configuration `AWS::Lambda::EventInvokeConfig`
You build this resource by building destinations, if they exist [on this line](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/compile/functions/index.js#L562)

I would check for `maxRetryAttempts` & `maxEventAgeInSeconds` on their own as part of the function configuration and create the destination even if no destinations were declared, is this ok? 

What happens if these are declared and they are re-declared on a `destination` config, should the `destination` config silently override the `function` config?

On a side note, should we accept them as global properties from provider (like `memorySize` and so on) or it doesn't make sense?

**Comment by medikoo** [comment]
@AhmedFat7y thanks for asking. Yes, we should probably rename `compileFunctionDestinations` into `compileEventInvokeConfig` and internally handle all options that trigger generation of that resource.

> What happens if these are declared and they are re-declared on a destination config, should the destination config silently override the function config?

Great question. I would do as follows:
- Favor setting on _function_ (not _destination_)
- Deprecate setting at _destination_ (so support it, if it is, but show deprecation log in such case) Unfortunately for that we need to implement `logDeprecation` first at https://github.com/serverless/serverless/issues/7422 - still that should not be difficult to do, and it's quite notable feature that will allow us to add more deprecations

> should we accept them as global properties from provider (like memorySize and so on) or it doesn't make sense?

I would not add it at this point. At least it applies only when function is invoked _asynchronously_, and accepting that globally will incur creation of `AWS::Lambda::EventInvokeConfig` for all functions, that may not be really wanted. Let's consider this only after we have a specific request for that.

**Comment by AhmedFat7y** [comment]
Will work on it

**Comment by AhmedFat7y** [comment]
I will start working on this given the deprecation flow is ready

**Comment by SanjayS5** [comment]
@AhmedFat7y Thanks  for working on this! How is it coming along? No rush of course, just curious :)

**Comment by AhmedFat7y** [comment]
@SanjayS5 I think I can open a PR on 27th of June

**Comment by pgrzesik** [comment]
Hello there :wave: I've noticed that as of right now, SQS doesn't support `maximumEventAge`, in a similar way as it supports `maximumRetryAttempts` - does it make sense to add support for that, linked to this issue or should it become a separate issue? Or in general, does it make sense to add it at all or it's going to be deprecated after this issue will be resolved?

**Comment by AhmedFat7y** [comment]
I’m not sure
I was unable to allocate time for this at all and I don’t think I will be able to any to anytime soon

**Comment by medikoo** [comment]
Th

**Comment by medikoo** [comment]
> I was unable to allocate time for this at all and I don’t think I will be able to any to anytime soon

@AhmedFat7y thanks for info, I've removed the assignment

> I've noticed that as of right now, SQS doesn't support `maximumEventAge`, in a similar way as it supports `maximumRetryAttempts`

@pgrzesik can you share more details?

**Comment by pgrzesik** [comment]
@medikoo Sure thing, I've read the first message again and I realized that it's not clear what I was trying to say - sorry about that! I realized that SQS doesn't really support `maximumRetryAttempts` as it's only for `streams`, similar to `maximumRecordAgeInSeconds` so please disregard my previous message. 

As for this issue, if I understand correctly, it should add the support for configuring `maxEventAge` directly on function and override any potential settings on destinations (`EventInvokeConfig` resources). Is there anything else that I should know before picking this up?

**Comment by medikoo** [comment]
@pgrzesik thanks for asking, I think we should support `maxEventAgeInSeconds` and `maxRetryAttempts` for all events that support it.

> if I understand correctly, it should add the support for configuring maxEventAge directly on function and override any potential settings on destinations

Yes, exactly, I believe we should do as proposed in [above comment](https://github.com/serverless/serverless/issues/7678#issuecomment-626568269) @AhmedFat7y implemented handling of deprecation logs, so we're not blocked. See https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#configuring-deprecations on how to configure deprecations

**Comment by pgrzesik** [comment]
@medikoo Thank you for explanation :bow: 

I would love to give it a shot if it's up for grabs :raised_hands:

**Comment by medikoo** [comment]
@pgrzesik totally take it. Thank you!

**Comment by pgrzesik** [comment]
Hello :wave: I started working on it and I feel like I'm either missing some piece of information or I misunderstood something here and would love to clarify that.

In the above comment by @AhmedFat7y (https://github.com/serverless/serverless/issues/7678#issuecomment-626568269), there's a suggestion: 
> I would check for maxRetryAttempts & maxEventAgeInSeconds on their own as part of the function configuration and create the destination even if no destinations were declared

First question that pops up to my mind is - what is the reason for creating it (`EventInvokeConfig`), if destinations are not declared? If I'm understanding it correctly, either `onSuccess` or `onFailure` is mandatory, so if they're not declared, what should be the default here? 

The section question depends on the answer to the above question and is based on assumption that we shouldn't create `EventInvokeConfig` if the destinations are not declared. If in order to create `EventInvokeConfig` we have to specify `destinations` on function, what is the benefit of moving `maxEventAge` configuration from `destinations` to `function` level?

@medikoo Would you be able to clarify the above a bit for me? I would really appreciate it :bow: 

Sorry for a bit lengthy message and for committing to an issue before fully understanding the scope

**Comment by medikoo** [comment]
> @medikoo Would you be able to clarify the above a bit for me? I would really appreciate it 🙇

Sure, no problem!

>  If I'm understanding it correctly, either onSuccess or onFailure is mandatory

I don't think it's the case. AFAIK you can successfully invoke lambdas asynchronously without any destinations being configured, and in such case all other `EventInvokeConfig` settings are effective.

> , so if they're not declared, what should be the default here?

I don't think they need defaults, `DestinationConfig` section is optional in CloudFormation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventinvokeconfig.html#cfn-lambda-eventinvokeconfig-destinationconfig

**Comment by pgrzesik** [comment]
@medikoo Thanks a lot for your explanation, that helps a ton - I should have the final PR ready pretty soon :+1:

**Comment by dcheckoway** [comment]
Thank you for working on this enhancement!  Can't wait to have a way for Serverless to control these:
<img width="872" alt="Screen Shot 2020-07-28 at 9 37 16 AM" src="https://user-images.githubusercontent.com/281311/88673399-8f98c300-d0b6-11ea-8830-9ce52c66a1a8.png">

The PR looks great!

**Comment by SanjayS5** [comment]
Hey there! Just tried this out and it doesn't appear to change the maximum age of event from the default of 6 hours. 

I tried setting it at the provider and the function level. Both did not work. Is it working for you guys?

**Comment by medikoo** [comment]
> I still cannot use this setting at the provider level to blanket all functions. Is that the intended functionality?

This setting applies only to functions intended to be invoked asynchronously, and (currently) there are no means to distinct between functions invoked synchronously and asynchronously, that's the reason there's no _global_ approach.

Still if it's not harmful to set this also for functions never intended to be invoked asynchronously, then it shouldn't be harmful to add that. Please open a new request if feel that's really valuable.
